### ISSUE TITLE
"/usr/bin/bash: conda: command not found" when running the blobtoolkit.smk metapipeline

### ISSUE THREAD

**Issue by git-hobo** [blobtoolkit/blobtoolkit#196]
# TL;DR:
Are the following elements of BlobToolKit compatible with each other, when trying to run the whole pipeline?
**Docker Container:**
[genomehubs/blobtoolkit:latest](https://hub.docker.com/r/genomehubs/blobtoolkit)

**Snakemake Pipeline Files:**
[blobtoolkit](https://github.com/blobtoolkit/blobtoolkit/tree/main)/[src](https://github.com/blobtoolkit/blobtoolkit/tree/main/src)/[blobtoolkit-pipeline](https://github.com/blobtoolkit/blobtoolkit/tree/main/src/blobtoolkit-pipeline)/[src](https://github.com/blobtoolkit/blobtoolkit/tree/main/src/blobtoolkit-pipeline/src)

**Tutorial:**
[pipeline-tutorials/running-the-pipeline-in-a-container](https://blobtoolkit.genomehubs.org/pipeline/pipeline-tutorials/running-the-pipeline-in-a-container/)

Am I missing something?

# Long Version:
I want to perform taxonomic partitioning with BlobToolKit on self made assemblies and I'm struggling to make sense of the BlobToolKit components between the genomehubs.org tutorials, the partially archived github repository documentation and the video tutorials.

I am currently testing BlobToolKit locally, but want to eventually move it to an HPC, which does not allow using Docker, so I am running the container with Singularity/Apptainer.
**Setup**:
I am invoking the container like this:
```
SNK_FILE=${HDD}/mites/mite_blast/blobtoolkit/src/blobtoolkit-pipeline/src/blobtoolkit-snakefiles/blobtoolkit.smk
SIF_FILE=${HDD}/sif/blobtoolkit.sif
WORKDIR=${HDD}/mites/mite_blast/BlobTrial
ASSEMBLY=MiteTestsample

BUSCO_DIR=${HDD}/mites/mite_blast/dbs/busco
UNIPROT_DIR=${HDD}/mites/mite_blast/dbs/uniprot
NT_DIR=${HDD}/mites/mite_blast/dbs/ncbi
TAXDUMP_DIR=${HDD}/mites/mite_blast/dbs/ncbi_taxdump
CONDA_DIR=${HDD}/conda/


THREADS=19

apptainer exec \
    -B $WORKDIR:/blobtoolkit/data \
    -B $BUSCO_DIR:/blobtoolkit/databases/busco \
    -B $UNIPROT_DIR:/blobtoolkit/databases/uniprot_db \
    -B $NT_DIR:/blobtoolkit/databases/ncbi_db \
    -B $TAXDUMP_DIR:/blobtoolkit/databases/ncbi_taxdump \
    -B $CONDA_DIR:/blobtoolkit/.conda \
    $SIF_FILE \
    snakemake -p \
        --use-conda \
        --conda-prefix /blobtoolkit/.conda \
        --directory /blobtoolkit/data \
        --configfile /blobtoolkit/data/$ASSEMBLY.yaml \
        --stats $ASSEMBLY.snakemake.stats \
        -j $THREADS \
        -s $SNK_FILE \
        --resources btk=1
```
with $HDD being defined above and my $WORKDIR having this content:
![image](https://github.com/blobtoolkit/blobtoolkit/assets/28047373/f471cd3d-deda-42c8-8525-de3caa026ae7)
For accessing the snakefiles I just cloned the current github repository.

**Error:**
When executing the commands above I get this message:
```
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 19
Rules claiming more threads will be scaled down.
Provided resources: btk=1
Job stats:
job                 count    min threads    max threads
----------------  -------  -------------  -------------
all                     1              1              1
run_sub_pipeline       11              1             19
total                  12              1             19

Select jobs to execute...

[Tue Oct 31 10:41:47 2023]
rule run_sub_pipeline:
    output: /blobtoolkit/windowmasker.stats
    log: logs/windowmasker/run_sub_pipeline.log
    jobid: 5
    benchmark: logs/windowmasker/run_sub_pipeline.benchmark.txt
    reason: Missing output files: /blobtoolkit/windowmasker.stats
    wildcards: tool=windowmasker
    resources: tmpdir=/tmp

snakemake -p           
-j 1           
--directory /blobtoolkit/windowmasker           
--configfile /blobtoolkit/config.yaml           
--latency-wait 60           
--stats /blobtoolkit/windowmasker.stats           
--restart-times 0           
-s hdd/mites/mite_blast/blobtoolkit/src/blobtoolkit-pipeline/src/blobtoolkit-snakefiles/windowmasker.smk 2> logs/windowmasker/run_sub_pipeline.log

/usr/bin/bash: conda: command not found
```
Now I guess the version of the pipeline, the version of the container and the version of the tutorial are just not compatible with each other. I have noticed a few things:
1. ```snakemake``` is right. ```conda``` is not available in the container, regardless if I run it interactively as a ```docker``` or ```singularity```/```apptainer``` container.
2. The (virtual?) directory ```/blobtoolkit/.conda``` is not available in the container unless you bind it, as opposed to the other directories like ```/blobtoolkit/data ``` which exist either way.
3. Invoking ```snakemake``` in the container with a dry run option throws no error messages.

I would appreciate any hints that may get me back on the right track.

**Comment by rjchallis** [comment]
Sorry for the confusing state of the docs. The simplest way to run the pipeline is with Docker or Singularity using the startup script in the `genomehubs/blobtoolkit` Docker image. This image should have all the files/dependencies in place to run the latest version of the pipeline.  

I've added a brief Docker based example to the end of the README in the develop branch at https://github.com/blobtoolkit/blobtoolkit/tree/develop#blobtoolkit-pipeline, but this should work similarly for singularity.

Hope this helps

**Comment by git-hobo** [comment]
Hello Richard,
thank you so much for answering!

Just in case any other singularity users come across this thread. I started the pipeline like:
```shell
singularity exec \
    -B ${DATASET_DIR}:/blobtoolkit/datasets \
    -B ${DATABASE_DIR}:/blobtoolkit/databases \
    -B ${OUTPUT}:/blobtoolkit/output \
    --env ASSEMBLY=$ASSEMBLY \
    --env THREADS=$THREADS \
    --env TRANSFER=$TRANSFER \
    $SIF_FILE startup.sh
```
with ```$SIF_FILE``` pointing to the downloaded Singularity Image File and the pipeline config file located at:
```shell
${DATASET_DIR}/${ASSEMBLY}/config.yaml
```
and this works for me.

**Comment by jsoghigian** [comment]
Hi @git-hobo  - This is a rather dumb question, but what is "startup.sh" in your case? I'm trying to adapt your solution to my cluster.

**Comment by git-hobo** [comment]
> Hi @git-hobo - This is a rather dumb question, but what is "startup.sh" in your case? I'm trying to adapt your solution to my cluster.

That's absolutely not a dumb question! I also did not know about this functionality of docker/singularity/apptainer before. You can execute scripts in the container that are "shipped" with the container and are only accessible/visible once the container is instantiated. What I mean is, if you run the container interactively like this:
```shell
$ singularity shell blobtoolkit.sif
$ cd /bloobtoolkit/
$ ls
```
you should be able to see the ```starup.sh``` in the listed files and directories:
![image](https://github.com/blobtoolkit/blobtoolkit/assets/28047373/25dbb923-c7b3-43a7-85e5-8299ed6df52e)
Just as a heads up: For me it was not smooth sailing after that point, there was a lot more weirdness and timeouts and depending on what the exact problem is, the relevant error logs can be buried quite deep in the created folder structures.
Good luck, sailor!

**Comment by jsoghigian** [comment]
Thanks @git-hobo  ... I actually was able to run most functionality adhoc following tutorials online and then learned about the pipeline, which is now helpful given genome sizes I'm working with... but if this proves too big a headache I'll just go back to piecemeal dataset construction by hand. Appreciate the tips!
